airbyte-cdk/bulk/core/load/src/test/kotlin/io/airbyte/cdk/state/MemoryManagerTest.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.state
+
+import io.micronaut.context.annotation.Replaces
+import io.micronaut.context.annotation.Requires
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest
+import jakarta.inject.Inject
+import jakarta.inject.Singleton
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.runTest
+import kotlinx.coroutines.withTimeout
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Test
+
+@MicronautTest
+class MemoryManagerTest {
+    @Inject lateinit var memoryManager: MemoryManager
+
+    @Singleton
+    @Replaces(MemoryManager::class)
+    @Requires(env = ["test"])
+    class MockAvailableMemoryProvider : AvailableMemoryProvider {
+        override val availableMemoryBytes: Long = 1000
+    }
+
+    @Test
+    fun testReserveBlocking() = runTest {
```

Comment:
we need tests imvolving several threads

Reply:
Yep, it took *a lot* but I did finally flush out the bug you were afraid of. See `testReserveBlockingMultithreaded() `

Reply:
took a lot of efforts, or a lot of jobs?

Reply:
A lot of jobs to get one error.

Reply:
yeah, I usually test those by having some king of test-triggered breakpoint in the production code. As much as I dislike having test-only code in production codepath, anything else is always going to be very flaky or involve a lot of jobs/threads/workers...